Doc/c-api/exceptions.rst

```diff
@@ -125,6 +125,44 @@ For convenience, some of these functions will always return a
 
    .. versionadded:: 3.5
 
+.. c:function:: void PyErr_ChainExceptions(PyObject *exc, PyObject *val, PyObject *tb)
+
+   Set the provided exception as the cause for the current exception. If there
+   is no exception currently set (:c:func:`PyErr_Occurred` returns ``0``) this
+   function behaves as like :c:func:`PyErr_Restore`, but if an exception is
+   already set, set the provided exception as the context (accessible from
+   Python through the :attr:`__context__` attribute of the exception) of the
+   one that is currently set.
```

Comment:
Isn't it the other way around - the existing exception becomes the context of the provided exception?

Reply:
The code calls `PyException_SetContext(val2, val);` where `val` is the provided one and `val2` is the one that is currently set. `PyException_SetContext` sets the second argument as the context for the first.

So unless I am missing something (is late here :sweat_smile: ) this means that "The provided exception (`val`) becomes the context of the existing exception (`val2`)"

Reply:
Right, I was thinking we are raising the provided exception in the context of the existing one, but that's what SetObject is for.   Can we think of a more descriptive name for this function? I confused the direction, and you confused cause/context because "chain" is ambiguous here.


How about PyErr_RestoreContext?   IIUC this describes what you use this function for - putting back a fetched exception, as the context of any exception that happened between the fetch and now.  (Should something like the previous sentence be in the doc?)

Reply:
Or possibly PyErr_RestoreAsContext?

Reply:
In general authors look for "chain" in the docs because that's the action they want to do, so I would really recommend to have "chain" in the name.

Reply:
That may be the case, but I think we should promote clear language, and "chain" in this context is vague. The doc can contain "chain" in the prose for the benefit of people who search for it (for now). But should it be part of the function name for all eternity?

Reply:
Is unclear to me. I agree with the ambiguity but what users if the function want is a higher level approach to the python behaviour: an exception is raised when another exception was there and I want to do in C the same thing as Python does.

I'm not a native speaker do I trust your judgement regarding how ambiguous this sounds and that we need a better name so let's search for one.

Maybe we should make a section in the docs about "exception chaining" and offer two APIs: one for cause and another for context

Reply:
@iritkatriel Maybe you can suggest a more specific name that still retains "chain" as part of it?

Reply:
The context is an implicitly chained exception (the cause is explicitly chained). There is no need to have both chained and context in the name. If we have only one of them I think it should be context and not chain.

Reply:
Given the equivalence  to `PyErr_Restore` when no exception is currently set, my public naming suggestion would be `PyErr_RestoreOrChain`, with the shorthand documentation being "Restores the given exception state, or chains it as the exception context if an exception is already set".

The key point to emphasise is that the function takes the same arguments as `PyErr_Restore` (retrieved from a prior `PyErr_Fetch` call), and generally does the same thing, but handles currently set exceptions differently: instead of clearing them, it leaves them in place, and "restores" the given exception as the exception context, rather than as the active exception.